okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/EquipPlan.kt

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (C) 2022 Block, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3.internal.connection
+
+import java.io.InterruptedIOException
+import java.net.ProtocolException
+import java.net.UnknownServiceException
+import java.security.cert.CertificateException
+import javax.net.ssl.SSLException
+import javax.net.ssl.SSLHandshakeException
+import javax.net.ssl.SSLPeerUnverifiedException
+import javax.net.ssl.SSLSocket
+import okhttp3.ConnectionSpec
+import okhttp3.Request
+import okio.IOException
+
+/**
+ * What to do once we have a socket but before we can make HTTP calls. This is usually a combination
+ * of the following features:
+ *
+ *  * CONNECT tunnels. When using an HTTP proxy to reach an HTTPS server we must send a CONNECT
+ *    request, and handle authorization challenges from the proxy.
+ *
+ *  * TLS handshakes.
+ *
+ * We might need many attempts at each of these steps. Tunnels fail due to authentication
+ * challenges, and TLS handshakes fail due to mismatched protocol versions. When we need another
+ * attempt, this class tracks what to try next.
+ *
+ * Unlike routes, we don't know that we'll need another equip plan until the current one fails.
+ */
+internal data class EquipPlan(
```

Comment:
Note that this replaces ConnectionSpec, which did the same thing but statefully. (Also ConnectionSpec didn’t know about tunnels.)